src/main/java/net/neoforged/neoforge/event/AddBlockToBlockEntityTypeEvent.java

```diff
@@ -0,0 +1,124 @@
+     */
+    public ResourceKey<BlockEntityType<?>> getBlockEntityTypeResourceKey() {
+        return blockEntityTypeResourceKey;
+    }
+
+    /**
+     * The current {@link BlockEntityType}.
+     */
+    public BlockEntityType<?> getBlockEntityType() {
+        return blockEntityType;
+    }
+
+    /**
+     * An immutable list of the list of valid blocks that will be assigned to replace {@link BlockEntityType#validBlocks}'s value.
+     */
+    public Set<Block> getCurrentValidBlocks() {
+        return Collections.unmodifiableSet(this.currentValidBlocks);
+    }
+
+    /**
+     * Will add the given block to the list of valid blocks we will assign to the current {@link BlockEntityType}.
+     * Make sure the given block is the same or derived from the highest common class of all the entries in valid blocks.
+     * <p></p>
+     * Example: If the valid blocks list has {@link StandingSignBlock} entry and {@link WallSignBlock} entry, the common class is SignBlock
+     * and the given block must be a {@link SignBlock} or has {@link SignBlock} as a child class in its hierarchy.
+     * 
+     * @param block The block to add as a valid block for the current {@link BlockEntityType}
+     */
+    public void addValidBlock(Block block) {
+        if (this.baseClass.get() == null || this.baseClass.get().isAssignableFrom(block.getClass())) {
+            this.currentValidBlocks.add(block);
+        } else {
+            throw new IllegalArgumentException("Given block " + block + " does not derive from existing valid block's common superclass of " + this.baseClass);
+        }
+    }
+
+    @Nullable
+    private Class<?> setCommonSuperClassForExistingValidBlocks() {
+        Class<?> calculatedBaseClass = null;
+
+        for (Block existingBlock : this.currentValidBlocks) {
+            if (calculatedBaseClass != null) {
+                calculatedBaseClass = findClosestCommonSuper(calculatedBaseClass, existingBlock.getClass());
+            } else {
+                calculatedBaseClass = existingBlock.getClass();
+            }
+        }
+
+        return calculatedBaseClass;
+    }
+
+    private static Class<?> findClosestCommonSuper(Class<?> a, Class<?> b) {
+        while (!a.isAssignableFrom(b)) {
+            a = a.getSuperclass();
+        }
+        return a;
+    }
+
+    @EventBusSubscriber(modid = "neoforge", bus = EventBusSubscriber.Bus.MOD)
+    private static class CommonHandler {
+        private static final MethodHandle METHOD_HANDLE;
+
+        static {
+            try {
+                METHOD_HANDLE = MethodHandles.privateLookupIn(BlockEntityType.class, MethodHandles.lookup()).findSetter(BlockEntityType.class, "validBlocks", Set.class);
+            } catch (NoSuchFieldException | IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        @SubscribeEvent
+        private static void onCommonSetup(FMLCommonSetupEvent event) throws Throwable {
```

Comment:
I would do this in `GameData.postRegisterEvents()` between the `SpawnPlacements.fireSpawnPlacementEvent()` and `CreativeModeTabRegistry.sortTabs()` calls